java-checks/src/main/java/org/sonar/java/checks/StandardFunctionalInterfaceCheck.java

```diff
@@ -0,0 +1,299 @@
+    return ImmutableList.of(Tree.Kind.INTERFACE);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    ClassTree classTree = (ClassTree) tree;
+    // classTree.simpleName() never null for Tree.Kind.INTERFACE
+    IdentifierTree issueLocation = classTree.simpleName();
+    Optional.of(classTree)
+      .filter(StandardFunctionalInterfaceCheck::isFunctionalInterface)
+      .filter(StandardFunctionalInterfaceCheck::isNonStandardFunctionalInterface)
+      .filter(StandardFunctionalInterfaceCheck::hasNoExtension)
+      .flatMap(StandardFunctionalInterfaceCheck::lookupFunctionalMethod)
+      .flatMap(StandardFunctionalInterfaceCheck::lookupMatchingStandardInterface)
+      .ifPresent(standardInterface -> reportIssue(issueLocation, buildIssueMessage(classTree, standardInterface)));
+  }
+
+  private static String buildIssueMessage(ClassTree interfaceTree, String standardInterface) {
+    if (interfaceTree.members().size() <= 1) {
+      return "Drop this interface in favor of \"" + standardInterface + "\".";
+    } else {
+      return "Make this interface extend \"" + standardInterface + "\" and remove the functional method declaration.";
+    }
+  }
+
+  private static Optional<MethodSymbol> lookupFunctionalMethod(ClassTree interfaceTree) {
+    return interfaceTree.symbol().memberSymbols().stream()
+      .filter(Symbol::isMethodSymbol)
+      .map(MethodSymbol.class::cast)
+      .filter(StandardFunctionalInterfaceCheck::isNonStaticNonDefaultMethod)
+      .filter(StandardFunctionalInterfaceCheck::isNotObjectMethod)
+      .findFirst();
+  }
+
+  private static Optional<String> lookupMatchingStandardInterface(MethodSymbol functionalMethod) {
+    MethodTree declaration = functionalMethod.declaration();
+    if (!functionalMethod.thrownTypes().isEmpty() || (declaration != null && !declaration.typeParameters().isEmpty())) {
+      return Optional.empty();
+    }
+    return STD_INTERFACE_BY_PARAMETER_COUNT.getOrDefault(functionalMethod.parameterTypes().size(), Collections.emptyList()).stream()
+      .map(standardInterface -> standardInterface.matchingSpecialization(functionalMethod))
+      .filter(specialization -> specialization != null)
+      .findFirst();
+  }
+
+  private static boolean hasNoExtension(ClassTree tree) {
+    return tree.superInterfaces().isEmpty();
+  }
+
+  private static boolean isFunctionalInterface(ClassTree tree) {
+    return tree.symbol().metadata().isAnnotatedWith("java.lang.FunctionalInterface");
+  }
+
+  private static boolean isNonStandardFunctionalInterface(ClassTree tree) {
+    return !STD_INTERFACE_NAMES.contains(tree.symbol().type().fullyQualifiedName());
+  }
+
+  private static boolean isNotObjectMethod(MethodSymbol method) {
+    MethodTree declaration = method.declaration();
+    return declaration == null || !OBJECT_METHODS.anyMatch(declaration);
+  }
+
+  private static boolean isNonStaticNonDefaultMethod(MethodSymbol methodSymbol) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused private "isNonStaticNonDefaultMethod" method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUnusedPrivateMethod)